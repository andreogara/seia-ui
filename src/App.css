.App {
  text-align: center;
  height: 100vh;
  display: flex;
  align-items: center;
  justify-content: center;
  background-color:  #282c34;
}

.App-logo {
  height: 40vmin;
  pointer-events: none;
}

@media (prefers-reduced-motion: no-preference) {
  .App-logo {
    animation: App-logo-spin infinite 20s linear;
  }
}

.App-link {
  color: #61dafb;
}

@keyframes App-logo-spin {
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
}

fieldset {
  padding: 30px;
  border: 3px solid rgb(104, 102, 102);
  text-align: center;
  font-size: 2ch;
  width: 40%;
}

legend {
  font-size: 2ch;
  font-family: serif;
}

form {
  padding-top: 10px;
  width: 100%;
  font-size: 1.2ch;
  background: white;
  height: 90%;
  border: 2px solid #282c34;
  overflow: scroll;
  text-align: -webkit-center;
}

fieldset label {
  margin-right: 5px;
  font-family: serif;
}

.summary-items {
  border-collapse: collapse;
  margin: 25px 0;
  font-size: 0.9em;
  font-family: sans-serif;
  min-width: 700px;
  box-shadow: 0 0 20px rgba(0, 0, 0, 0.15);
  margin-left: 50px;
}

.summary-items thead tr {
  background-color: #282c34;
  color: #ffffff;
  text-align: left;
}
.summary-items caption {
  font-family: serif;
  font-size: 3ch;
}

.summary-items th,
.summary-items td {
    padding: 12px 15px;
    text-align: center;
}

.summary-items tbody tr {
  border-bottom: 1px solid #dddddd;
}

.summary-items tbody tr:nth-of-type(even) {
  background-color: #f3f3f3;
}

.summary-items tbody tr:last-of-type {
  border-bottom: 2px solid #282c34;
}

.generate-map-button, .save-config-button, .reset-estimates-button {
  align-items: center;
  background: #FFFFFF;
  border: 0 solid #E2E8F0;
  box-shadow: 0 4px 6px -1px rgba(0, 0, 0, 0.1), 0 2px 4px -1px rgba(0, 0, 0, 0.06);
  box-sizing: border-box;
  color: #1A202C;
  display: inline-flex;
  font-family: Inter, sans-serif;
  font-size: 1rem;
  font-weight: 700;
  height: 56px;
  justify-content: center;
  line-height: 24px;
  padding: 24px;
  text-decoration: none;
  width: auto;
  border-radius: 8px;
  cursor: pointer;
  margin-right: 5px;
}

.site-equipment-layout {
  border: 2px solid rgb(24, 23, 23);
  position: fixed;
  background-color: #e7e7e7;
  left: 40%;
  top: 20%;
  overflow: auto;
}

.site-equipment-layout > button {
  float: right;
  background-color: transparent;
  font-size: 8px;
  border: none;
}

.site-equipment {
  clear: both;
  display: inline-block;
  margin: 7px 10px;
  background-size:cover;
  background-repeat: no-repeat;
  border: 1px solid black;
}