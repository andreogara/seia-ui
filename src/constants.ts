export type BatteryConfigMap = {
    name: string;
    dimension: {
        length: number;
        width: number;
    }
    energy: number;
    cost: number;
    release?: number;
}

export type ConfigState = {
    [key: string]: number;
}

export const EquipmentColorCoding:{[key:string]: string} = {
    transformer: "red",
    megapackxl: "blue",
    megapack2: "cyan",
    megapack: "orange",
    powerpack: "yellow"
}

export const batteryConfigurationMap = [
    {
        name: "megapackxl",
        dimension: {
            length: 40,
            width: 10
        },
        energy: 4,
        cost: 120000,
        release: 2022
    },
    {
        name: "megapack2",
        dimension: {
            length: 30,
            width: 10
        },
        energy: 3,
        cost: 80000,
        release: 2021
    },
    {
        name: " megapack",
        dimension: {
            length: 30,
            width: 10
        },
        energy: 2,
        cost: 50000,
        release: 2005
    },
    {
        name: "powerpack",
        dimension: {
            length: 10,
            width: 10
        },
        energy: 1,
        cost: 10000,
        release: 2000
    }
]

export const TransformerConfig = {
    name: "transformer",
    dimension: {
        length: 10,
        width: 10
    },
    energy: -0.5,
    cost: 10000
}

export const metrics = {
    distance: "FT",
    energy: "MWh"
}