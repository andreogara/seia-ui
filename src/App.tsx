import React from 'react';
import './App.css';
import { batteryConfigurationMap, TransformerConfig, metrics, ConfigState, BatteryConfigMap } from './constants';
import { TableRowItem } from './TableRowItem';
import { SiteEquipmentLayout } from './SiteEquipmentLayout';
import { getConfigData, saveConfigData } from "./api";
function App() {
  const [configState, setConfigState] = React.useState<ConfigState>({});
  const [shouldShowSiteLayout, setshouldShowSiteLayout] = React.useState(false);

  React.useEffect(() => {
    console.log(getConfigData);
    getConfigData().then(data => {
      if(data) {
        setConfigState(JSON.parse(data));
      }
    })
  }, []);

  const generateMap = (e: React.SyntheticEvent<HTMLButtonElement>) => {
    e.preventDefault();
    setshouldShowSiteLayout(true);
  }

  const handleChange = (e: React.SyntheticEvent<HTMLInputElement>) => {
    const val = parseInt(e.currentTarget.value);
    const key = e.currentTarget.name;
    setConfigState(prevState => {
      const currentNumberOfAllBatteries = Object.keys(configState).reduce((sum, currItem) => {
        if (currItem !== TransformerConfig.name) {
          return sum + configState[currItem];
        }
        else {
          return sum;
        }
      }, 0);
      const newTransformersCount = Math.ceil(currentNumberOfAllBatteries / 2);
      return {
        ...prevState,
        [key]: val,
        [TransformerConfig.name]: newTransformersCount
      }
    });
  }

  const saveConfig = (e: React.SyntheticEvent<HTMLButtonElement>) => {
    e.preventDefault();
    saveConfigData(configState)
      .then(() => alert("sucessfully saved!"))
      .catch(err => console.log(err));
  }

  const getBatteryConfigTemplate = (configurationMap: BatteryConfigMap[]) => {
    const result = [];
    for (let config of configurationMap) {
      result.push((
        <p key={config.name}>
          <label htmlFor={config.name}>{config.name.toUpperCase()}</label>
          <input type="number" min={0} id={config.name} name={config.name} onChange={handleChange} value={configState[config.name] || 0} />
        </p>
      ))
    }

    return result;
  }

  const convertToDollarCurrency = (value: number) => {
    return value.toLocaleString('en-US', {
      style: "currency",
      currency: 'USD'
    });
  }

  let totalSquareFootageRequired = 0;
  let totalEnergyRequirements = 0;
  let totalCostsRequired = 0;
  let totalNumberOfAllEquipments = 0;

  const buildTableOfEstimates = () => {
    const summaryItems = [];
    for (const item in configState) {
      const numberOfSpecificEquipment = configState[item] as number;
      const itemDetails = batteryConfigurationMap.find(config => config.name === item);
      const costOfItems = itemDetails?.cost || 0;
      let totalCostsForSpecificItem = costOfItems * numberOfSpecificEquipment;
      let totalEnergyCostsForSpecificItem = itemDetails?.energy ? itemDetails?.energy * numberOfSpecificEquipment : 0;
      let totalSquareFootageForSpecificItem = itemDetails?.dimension ? (itemDetails.dimension.length * itemDetails.dimension.width) * numberOfSpecificEquipment : 0;
      if (item === TransformerConfig.name) {
        totalCostsForSpecificItem = TransformerConfig.cost * numberOfSpecificEquipment;
        totalEnergyCostsForSpecificItem = TransformerConfig.energy * numberOfSpecificEquipment;
        totalSquareFootageForSpecificItem = Math.pow(TransformerConfig.dimension.length, 2) * numberOfSpecificEquipment;
      }
      const totalCostInLocalCurrency = convertToDollarCurrency(totalCostsForSpecificItem);
      totalCostsRequired += totalCostsForSpecificItem;
      totalEnergyRequirements += totalEnergyCostsForSpecificItem;
      totalSquareFootageRequired += totalSquareFootageForSpecificItem;
      totalNumberOfAllEquipments += numberOfSpecificEquipment;
      summaryItems.push(
        <TableRowItem
          key={item}
          itemName={item}
          itemCount={numberOfSpecificEquipment}
          totalEnergyCostsForSpecificItem={`${totalEnergyCostsForSpecificItem}${metrics.energy}`}
          totalSquareFootageForSpecificItem={`${totalSquareFootageForSpecificItem}${metrics.distance}`}
          totalCostInLocalCurrency={totalCostInLocalCurrency}
        />
      )
    }
    return summaryItems;
  }

  return (
    <div className="App">
      <form>
        <fieldset disabled={shouldShowSiteLayout}>
          <legend>Choose your battery configuration</legend>
          {getBatteryConfigTemplate(batteryConfigurationMap)}
          <button className="reset-estimates-button" type='button' onClick={() => setConfigState({})}>Reset fields</button>
        </fieldset>

        <table id="config-summary" className='summary-items'>
          <caption>Site Equipment Estimates</caption>
          <thead>
            <tr>
              <th>Name</th>
              <th>Count</th>
              <th>Cost</th>
              <th>Capacity</th>
              <th>Square Footage</th>
            </tr>
          </thead>
          <tbody>
            {buildTableOfEstimates()}
          </tbody>
          <tfoot>

            <tr>
              <td>Total</td>
              <td>
                {totalNumberOfAllEquipments}
              </td>
              <td>
                {convertToDollarCurrency(totalCostsRequired)}
              </td>
              <td>
                {`${totalEnergyRequirements}${metrics.energy}`}
              </td>
              <td>
                {`${totalSquareFootageRequired}${metrics.distance}`}
              </td>
            </tr>
          </tfoot>
        </table>
        <button className="generate-map-button" onClick={generateMap}>Generate Site Map</button>
        <button className="save-config-button" onClick={saveConfig}>Save Config</button>
        {shouldShowSiteLayout && <SiteEquipmentLayout configState={configState} hide={() => setshouldShowSiteLayout(false)} />}
      </form>
    </div>
  );
}

export default App;
