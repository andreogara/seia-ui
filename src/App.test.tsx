import React from 'react';
import { render, screen, within } from '@testing-library/react';
import App from './App';
import { batteryConfigurationMap} from './constants';
import userEvent from '@testing-library/user-event';

test('renders the battery selectors correctly', () => {
  render(<App />);
  const batterySelectors = screen.getByRole("group");
  expect(batterySelectors).toMatchSnapshot();
});

test('updates table of estimates correctly when one selects a battery amount', () => {
  render(<App />);
  const megapackXL = screen.getByLabelText(batteryConfigurationMap[0].name.toUpperCase());
  userEvent.type(megapackXL, "5");
  const table = screen.getByRole("table");
  const estimates = within(table).getAllByRole("row");
  expect(estimates).toMatchSnapshot();
});

test('updates table of estimates correctly when one selects a battery amount', () => {
  render(<App />);
  const megapackXL = screen.getByLabelText(batteryConfigurationMap[0].name.toUpperCase());
  userEvent.type(megapackXL, "5");
  const table = screen.getByRole("table");
  const estimates = within(table).getAllByRole("row");
  expect(estimates).toMatchSnapshot();
});

test('clears the estimate button when reset button is clicked', () => {
  render(<App />);
  const clearButton = screen.getByRole("button", {name: "Reset fields"});
  userEvent.click(clearButton);
  const table = screen.getByRole("table");
  expect(table).toMatchSnapshot();
});

test('generates UI layout depending on selected configuration when appropriate button is clicked', async () => {
  render(<App />);
  const siteMapButton = screen.getByRole("button", {name: "Generate Site Map"});
  userEvent.click(siteMapButton);
  const layout = await screen.findByRole("region", {name: "Site equipment layout"});
  expect(layout).toBeInTheDocument();
});
