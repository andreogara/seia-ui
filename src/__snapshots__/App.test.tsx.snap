// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`clears the estimate button when reset button is clicked 1`] = `
<table
  class="summary-items"
  id="config-summary"
>
  <caption>
    Site Equipment Estimates
  </caption>
  <thead>
    <tr>
      <th>
        Name
      </th>
      <th>
        Count
      </th>
      <th>
        Cost
      </th>
      <th>
        Capacity
      </th>
      <th>
        Square Footage
      </th>
    </tr>
  </thead>
  <tbody />
  <tfoot>
    <tr>
      <td>
        Total
      </td>
      <td>
        0
      </td>
      <td>
        $0.00
      </td>
      <td>
        0MWh
      </td>
      <td>
        0FT
      </td>
    </tr>
  </tfoot>
</table>
`;

exports[`renders the battery selectors correctly 1`] = `
<fieldset>
  <legend>
    Choose your battery configuration
  </legend>
  <p>
    <label
      for="megapackxl"
    >
      MEGAPACKXL
    </label>
    <input
      id="megapackxl"
      min="0"
      name="megapackxl"
      type="number"
      value="0"
    />
  </p>
  <p>
    <label
      for="megapack2"
    >
      MEGAPACK2
    </label>
    <input
      id="megapack2"
      min="0"
      name="megapack2"
      type="number"
      value="0"
    />
  </p>
  <p>
    <label
      for=" megapack"
    >
       MEGAPACK
    </label>
    <input
      id=" megapack"
      min="0"
      name=" megapack"
      type="number"
      value="0"
    />
  </p>
  <p>
    <label
      for="powerpack"
    >
      POWERPACK
    </label>
    <input
      id="powerpack"
      min="0"
      name="powerpack"
      type="number"
      value="0"
    />
  </p>
  <button
    type="button"
  >
    Reset fields
  </button>
</fieldset>
`;

exports[`updates table of estimates correctly when one selects a battery amount 1`] = `
Array [
  <tr>
    <th>
      Name
    </th>
    <th>
      Count
    </th>
    <th>
      Cost
    </th>
    <th>
      Capacity
    </th>
    <th>
      Square Footage
    </th>
  </tr>,
  <tr>
    <td>
      megapackxl
    </td>
    <td>
      5
    </td>
    <td>
      $600,000.00
    </td>
    <td>
      20MWh
    </td>
    <td>
      2000FT
    </td>
  </tr>,
  <tr>
    <td>
      transformer
    </td>
    <td>
      0
    </td>
    <td>
      $0.00
    </td>
    <td>
      0MWh
    </td>
    <td>
      0FT
    </td>
  </tr>,
  <tr>
    <td>
      Total
    </td>
    <td>
      5
    </td>
    <td>
      $600,000.00
    </td>
    <td>
      20MWh
    </td>
    <td>
      2000FT
    </td>
  </tr>,
]
`;

exports[`updates table of estimates correctly when one selects a battery amount 2`] = `
Array [
  <tr>
    <th>
      Name
    </th>
    <th>
      Count
    </th>
    <th>
      Cost
    </th>
    <th>
      Capacity
    </th>
    <th>
      Square Footage
    </th>
  </tr>,
  <tr>
    <td>
      megapackxl
    </td>
    <td>
      5
    </td>
    <td>
      $600,000.00
    </td>
    <td>
      20MWh
    </td>
    <td>
      2000FT
    </td>
  </tr>,
  <tr>
    <td>
      transformer
    </td>
    <td>
      0
    </td>
    <td>
      $0.00
    </td>
    <td>
      0MWh
    </td>
    <td>
      0FT
    </td>
  </tr>,
  <tr>
    <td>
      Total
    </td>
    <td>
      5
    </td>
    <td>
      $600,000.00
    </td>
    <td>
      20MWh
    </td>
    <td>
      2000FT
    </td>
  </tr>,
]
`;
